--- a/unix/mpconfigport.mk
+++ b/unix/mpconfigport.mk
@@ -34,6 +34,19 @@ MICROPY_SSL_MBEDTLS = 0
 # jni module requires JVM/JNI
 MICROPY_PY_JNI = 0
 
+MICROPY_PY_TREZORCONFIG = 1
+
+MICROPY_PY_TREZORCRYPTO = 1
+
+MICROPY_PY_TREZORDEBUG = 1
+
+MICROPY_PY_TREZORMSG = 1
+
+# TrezorUi module requires SDL2+SDL2_image unless TREZORUI_NOUI=1 is used
+MICROPY_PY_TREZORUI = 1
+
+MICROPY_PY_TREZORUTILS = 1
+
 # Avoid using system libraries, use copies bundled with MicroPython
 # as submodules (currently affects only libffi).
 MICROPY_STANDALONE = 0
