--- a/unix/mpconfigport.h
+++ b/unix/mpconfigport.h
@@ -131,6 +131,12 @@
 #define MICROPY_PY_MACHINE_PULSE    (1)
 #define MICROPY_MACHINE_MEM_GET_READ_ADDR   mod_machine_mem_get_addr
 #define MICROPY_MACHINE_MEM_GET_WRITE_ADDR  mod_machine_mem_get_addr
+#define MICROPY_PY_TREZORCONFIG     (1)
+#define MICROPY_PY_TREZORCRYPTO     (1)
+#define MICROPY_PY_TREZORDEBUG      (1)
+#define MICROPY_PY_TREZORMSG        (1)
+#define MICROPY_PY_TREZORUI         (1)
+#define MICROPY_PY_TREZORUTILS      (1)
 
 #define MICROPY_FATFS_ENABLE_LFN       (1)
 #define MICROPY_FATFS_RPATH            (2)
@@ -167,6 +173,12 @@ extern const struct _mp_obj_module_t mp_module_termios;
 extern const struct _mp_obj_module_t mp_module_socket;
 extern const struct _mp_obj_module_t mp_module_ffi;
 extern const struct _mp_obj_module_t mp_module_jni;
+extern const struct _mp_obj_module_t mp_module_TrezorConfig;
+extern const struct _mp_obj_module_t mp_module_TrezorCrypto;
+extern const struct _mp_obj_module_t mp_module_TrezorDebug;
+extern const struct _mp_obj_module_t mp_module_TrezorMsg;
+extern const struct _mp_obj_module_t mp_module_TrezorUi;
+extern const struct _mp_obj_module_t mp_module_TrezorUtils;
 
 #if MICROPY_PY_UOS_VFS
 #define MICROPY_PY_UOS_VFS_DEF { MP_ROM_QSTR(MP_QSTR_uos_vfs), MP_ROM_PTR(&mp_module_uos_vfs) },
@@ -203,6 +215,38 @@ extern const struct _mp_obj_module_t mp_module_jni;
 #else
 #define MICROPY_PY_USELECT_DEF
 #endif
+#if MICROPY_PY_TREZORCONFIG
+#define MICROPY_PY_TREZORCONFIG_DEF { MP_ROM_QSTR(MP_QSTR_TrezorConfig), MP_ROM_PTR(&mp_module_TrezorConfig) },
+#else
+#define MICROPY_PY_TREZORCONFIG_DEF
+#endif
+#if MICROPY_PY_TREZORCRYPTO
+#define MICROPY_PY_TREZORCRYPTO_DEF { MP_ROM_QSTR(MP_QSTR_TrezorCrypto), MP_ROM_PTR(&mp_module_TrezorCrypto) },
+#else
+#define MICROPY_PY_TREZORCRYPTO_DEF
+#endif
+#if MICROPY_PY_TREZORDEBUG
+#define MICROPY_PY_TREZORDEBUG_DEF { MP_ROM_QSTR(MP_QSTR_TrezorDebug), MP_ROM_PTR(&mp_module_TrezorDebug) },
+#else
+#define MICROPY_PY_TREZORDEBUGI_DEF
+#endif
+#if MICROPY_PY_TREZORMSG
+#define MICROPY_PY_TREZORMSG_DEF { MP_ROM_QSTR(MP_QSTR_TrezorMsg), MP_ROM_PTR(&mp_module_TrezorMsg) },
+#else
+#define MICROPY_PY_TREZORMSG_DEF
+#endif
+#if MICROPY_PY_TREZORUI
+#define MICROPY_PY_TREZORUI_DEF { MP_ROM_QSTR(MP_QSTR_TrezorUi), MP_ROM_PTR(&mp_module_TrezorUi) },
+#else
+#define MICROPY_PY_TREZORUI_DEF
+#endif
+#if MICROPY_PY_TREZORUTILS
+#define MICROPY_PY_TREZORUTILS_DEF { MP_ROM_QSTR(MP_QSTR_TrezorUtils), MP_ROM_PTR(&mp_module_TrezorUtils) },
+#else
+#define MICROPY_PY_TREZORUTILS_DEF
+#endif
+
+
 
 #define MICROPY_PORT_BUILTIN_MODULES \
     MICROPY_PY_FFI_DEF \
@@ -214,6 +258,12 @@ extern const struct _mp_obj_module_t mp_module_jni;
     MICROPY_PY_UOS_VFS_DEF \
     MICROPY_PY_USELECT_DEF \
     MICROPY_PY_TERMIOS_DEF \
+    MICROPY_PY_TREZORCONFIG_DEF \
+    MICROPY_PY_TREZORCRYPTO_DEF \
+    MICROPY_PY_TREZORDEBUG_DEF \
+    MICROPY_PY_TREZORMSG_DEF \
+    MICROPY_PY_TREZORUI_DEF \
+    MICROPY_PY_TREZORUTILS_DEF \
 
 // type definitions for the specific machine
 
